File: src/main/Timer.java
package main;

import javax.swing.SwingUtilities;

public class Timer extends Thread
{

	// time consumed by each frame
	public static int inter = 0;

	// time per frame, in ms
	public static final int TPF = 33;

	protected static int state;

	private static int delay = 0;

	private Inv thr = null;

	@Override
	public void run()
	{
		while (true)
		{
			long m = System.currentTimeMillis();
			state = 0;
			SwingUtilities.invokeLater(thr = new Inv());
			try
			{
				boolean end = false;
				while (!end)
				{
					synchronized (thr)
					{
						end = state == 1;
					}
					if (!end)
						sleep(1);
				}
				thr.join();
				delay = (int) (System.currentTimeMillis() - m);
				inter = (inter * 9 + 100 * delay / TPF) / 10;
				int sle = delay >= TPF ? 1 : TPF - delay;
				sleep(sle);
			}
			catch (InterruptedException e)
			{
				return;
			}
		}
	}

}

class Inv extends Thread
{
	@Override
	public void run()
	{
		MainBCU.timer(-1);
		synchronized (this)
		{
			Timer.state = 1;
		}
	}
}

File: src/main/Timer.kt
package main

import main.Timer
import page.MainFrame
import javax.swing.SwingUtilities

@Strictfp
class Timer : Thread() {
    private var thr: Inv? = null
    override fun run() {
        while (true) {
            val m = System.currentTimeMillis()
            Timer.Companion.state = 0
            SwingUtilities.invokeLater(Inv().also { thr = it })
            try {
                var end = false
                while (!end) {
                    synchronized(thr!!) { end = Timer.Companion.state == 1 }
                    if (!end) sleep(1)
                }
                thr!!.join()
                Timer.Companion.delay = (System.currentTimeMillis() - m).toInt()
                Timer.Companion.inter = (Timer.Companion.inter * 9 + 100 * Timer.Companion.delay / Timer.Companion.p) / 10
                val sle = if (Timer.Companion.delay >= Timer.Companion.p) 1 else Timer.Companion.p - Timer.Companion.delay
                sleep(sle.toLong())
            } catch (e: InterruptedException) {
                return
            }
        }
    }

    companion object {
        var p = 33
        var inter = 0
        var state = 0
        private const val delay = 0
    }
}

@Strictfp
internal class Inv : Thread() {
    override fun run() {
        MainFrame.Companion.timer(-1)
        synchronized(this) { Timer.Companion.state = 1 }
    }
}
